test_max_runtime_exceeded sometimes fails the reason code assertion

Example here:

https://travis-ci.org/twosigma/Cook/jobs/223179765#L1393-L1398

I can't seem to find any usages of reason `:mesos-command-executor-failed` (or 99003) in the scheduler. @wyegelwel, do you have any ideas as to what might cause this?

Saw this here again: https://travis-ci.org/twosigma/Cook/jobs/244625309
```
FAIL: test_max_runtime_exceeded (tests.cook.test_basic.CookTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/twosigma/Cook/integration/tests/cook/test_basic.py", line 63, in test_max_runtime_exceeded
    self.assertEqual(2003, job['instances'][0]['reason_code'])
AssertionError: 2003 != 99003
```

I got a different error (same test):
```
FAIL: test_max_runtime_exceeded (tests.cook.test_basic.CookTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/twosigma/Cook/integration/tests/cook/test_basic.py", line 222, in test_max_runtime_exceeded
    self.assertEqual(2003, job['instances'][0]['reason_code'])
AssertionError: 2003 != 6002
```
Reason code 6002 is REASON_EXECUTOR_UNREGISTERED.

Should we just mark this test explicit until someone has time to look into it?

Got this:

```bash
======================================================================
FAIL: test_max_runtime_exceeded (tests.cook.test_basic.CookTest)
----------------------------------------------------------------------
Traceback (most recent call last):
 File "/home/travis/build/twosigma/Cook/integration/tests/cook/test_basic.py", line 222, in test_max_runtime_exceeded
 self.assertEqual(2003, job['instances'][0]['reason_code'])
AssertionError: 2003 != 6002
-------------------- >> begin captured logging << --------------------
tests.cook.util: INFO: Using cook url http://localhost:12321
tests.cook.util: INFO: Using cook url http://localhost:12321
tests.cook.util: DEBUG: Waiting for connection to cook...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET / HTTP/1.1" 404 None
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /settings HTTP/1.1" 200 None
urllib3.connectionpool: DEBUG: http://172.17.0.4:5050 "GET /redirect HTTP/1.1" 307 0
tests.cook.util: INFO: Using mesos url http://172.17.0.4:5050
tests.cook.util: DEBUG: Waiting for connection to cook...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET / HTTP/1.1" 404 None
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /settings HTTP/1.1" 200 None
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /settings HTTP/1.1" 200 None
tests.cook.util: INFO: {'jobs': [{'command': 'sleep 135', 'cpus': 1, 'max_retries': 1, 'mem': 256, 'name': 'echo', 'priority': 1, 'uuid': '45963c35-6a35-4945-8d80-1f2a9fd8fd33', 'executor': 'cook', 'max_runtime': 5000}]}
urllib3.connectionpool: DEBUG: http://localhost:12321 "POST /rawscheduler HTTP/1.1" 201 None
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status waiting, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status waiting, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status waiting, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status running, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status running, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status running, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status running, expecting completed.
tests.cook.util: INFO: wait_until condition not yet met, retrying...
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
tests.cook.util: INFO: Job 45963c35-6a35-4945-8d80-1f2a9fd8fd33 has status completed, expecting completed.
tests.cook.util: INFO: wait_until condition satisfied
urllib3.connectionpool: DEBUG: http://localhost:12321 "GET /rawscheduler?job=45963c35-6a35-4945-8d80-1f2a9fd8fd33 HTTP/1.1" 200 None
--------------------- >> end captured logging << ---------------------
```

--- a/integration/tests/cook/reasons.py
+++ b/integration/tests/cook/reasons.py
@@ -1,0 +1,6 @@
+# Named constants for failure reason codes from cook or mesos.
+# See scheduler/src/cook/mesos/schema.clj for the reason code names.
+
+MAX_RUNTIME_EXCEEDED = 2003
+EXECUTOR_UNREGISTERED = 6002
+CMD_NON_ZERO_EXIT = 99003

--- a/integration/tests/cook/test_basic.py
+++ b/integration/tests/cook/test_basic.py
@@ -223,22 +223,24 @@ class CookTest(unittest.TestCase):
             util.wait_for_job(self.cook_url, job_uuid, 'running')
             util.wait_for_job(self.cook_url, job_uuid, 'completed', job_timeout_interval_ms)
             job = util.wait_for_end_time(self.cook_url, job_uuid)
-            self.assertEqual(1, len(job['instances']))
+            job_details = f"Job details: {json.dumps(job, sort_keys=True)}"
+            self.assertEqual(1, len(job['instances']), job_details)
             instance = job['instances'][0]
-            self.assertEqual('failed', instance['status'])
+            self.assertEqual('failed', instance['status'], job_details)
             valid_reasons = [# we killed the job because it took too long
                              CookTest.REASON_MAX_RUNTIME_EXCEEDED,
                              # the job was killed before the executor was properly registered
                              CookTest.REASON_EXECUTOR_UNREGISTERED,
                              # the job didn't die gracefully when we terminated it
                              CookTest.REASON_CMD_NON_ZERO_EXIT]
-            self.assertIn(instance['reason_code'], valid_reasons)
+            self.assertIn(instance['reason_code'], valid_reasons, job_details)
             actual_running_time_ms = instance['end_time'] - instance['start_time']
-            self.assertGreater(actual_running_time_ms, max_runtime_ms)
+            self.assertGreater(actual_running_time_ms, max_runtime_ms, job_details)
+            self.assertGreater(job_timeout_interval_ms, actual_running_time_ms, job_details)
             if job_executor_type == 'cook':
                 job = util.wait_for_exit_code(self.cook_url, job_uuid)
-                self.assertNotEqual(0, job['instances'][0]['exit_code'])
-                self.assertTrue(bool(job['instances'][0]['sandbox_directory']))
+                self.assertNotEqual(0, job['instances'][0]['exit_code'], job_details)
+                self.assertTrue(bool(job['instances'][0]['sandbox_directory']), job_details)
         finally:
             util.kill_jobs(self.cook_url, [job_uuid])
 

--- a/integration/tests/cook/test_basic.py
+++ b/integration/tests/cook/test_basic.py
@@ -208,26 +208,35 @@ class CookTest(unittest.TestCase):
                                                         'timeout-interval-minutes')
         # the value needs to be a little more than 2 times settings_timeout_interval_minutes to allow
         # at least two runs of the lingering task killer
-        job_timeout_interval_seconds = (2 * settings_timeout_interval_minutes * 60) + 15
-        job_timeout_interval_ms = job_timeout_interval_seconds * 1000
+        job_sleep_seconds = (2 * settings_timeout_interval_minutes * 60) + 15
+        job_sleep_ms = job_sleep_seconds * 1000
         max_runtime_ms = 5000
-        assert max_runtime_ms < job_timeout_interval_ms
+        assert max_runtime_ms < job_sleep_ms
         try:
+            # schedule a job that will go over its max_runtime time limit
             job_uuid, resp = util.submit_job(self.cook_url,
-                                             command=f'sleep {job_timeout_interval_seconds}',
+                                             command=f'sleep {job_sleep_seconds}',
                                              executor=job_executor_type, max_runtime=max_runtime_ms)
             self.assertEqual(201, resp.status_code, msg=resp.content)
+            # We wait for the job to start running, and only then start waiting for the max-runtime timeout,
+            # otherwise we could get a false-negative on wait-for-job 'completed' because of a scheduling delay.
+            # We wait for the 'end_time' attribute separately because there is another small delay
+            # between the job being killed and that attribute being set.
+            # Having three separate waits also disambiguates the root cause of a wait-timeout failure.
             util.wait_for_job(self.cook_url, job_uuid, 'running')
-            util.wait_for_job(self.cook_url, job_uuid, 'completed', job_timeout_interval_ms)
+            util.wait_for_job(self.cook_url, job_uuid, 'completed', job_sleep_ms)
             job = util.wait_for_end_time(self.cook_url, job_uuid)
             job_details = f"Job details: {json.dumps(job, sort_keys=True)}"
             self.assertEqual(1, len(job['instances']), job_details)
             instance = job['instances'][0]
+            # did the job fail as expected?
             self.assertEqual('failed', instance['status'], job_details)
             self.assertEqual(instance['reason_code'], reasons.MAX_RUNTIME_EXCEEDED, job_details)
+            # was the actual running time consistent with running over time and being killed?
             actual_running_time_ms = instance['end_time'] - instance['start_time']
             self.assertGreater(actual_running_time_ms, max_runtime_ms, job_details)
-            self.assertGreater(job_timeout_interval_ms, actual_running_time_ms, job_details)
+            self.assertGreater(job_sleep_ms, actual_running_time_ms, job_details)
+            # verify additional fields set when the cook executor is used
             if job_executor_type == 'cook':
                 job = util.wait_for_exit_code(self.cook_url, job_uuid)
                 self.assertNotEqual(0, job['instances'][0]['exit_code'], job_details)

--- a/integration/tests/cook/test_basic.py
+++ b/integration/tests/cook/test_basic.py
@@ -8,14 +8,11 @@ import uuid
 from nose.plugins.attrib import attr
 from retrying import retry
 
+from tests.cook import reasons
 from tests.cook import util
 
 class CookTest(unittest.TestCase):
     _multiprocess_can_split_ = True
-
-    REASON_CMD_NON_ZERO_EXIT     = 99003
-    REASON_MAX_RUNTIME_EXCEEDED  =  2003
-    REASON_EXECUTOR_UNREGISTERED =  6002
 
     @staticmethod
     def minimal_group(**kwargs):
@@ -227,13 +224,7 @@ class CookTest(unittest.TestCase):
             self.assertEqual(1, len(job['instances']), job_details)
             instance = job['instances'][0]
             self.assertEqual('failed', instance['status'], job_details)
-            valid_reasons = [# we killed the job because it took too long
-                             CookTest.REASON_MAX_RUNTIME_EXCEEDED,
-                             # the job was killed before the executor was properly registered
-                             CookTest.REASON_EXECUTOR_UNREGISTERED,
-                             # the job didn't die gracefully when we terminated it
-                             CookTest.REASON_CMD_NON_ZERO_EXIT]
-            self.assertIn(instance['reason_code'], valid_reasons, job_details)
+            self.assertEqual(instance['reason_code'], reasons.MAX_RUNTIME_EXCEEDED, job_details)
             actual_running_time_ms = instance['end_time'] - instance['start_time']
             self.assertGreater(actual_running_time_ms, max_runtime_ms, job_details)
             self.assertGreater(job_timeout_interval_ms, actual_running_time_ms, job_details)
@@ -807,7 +798,7 @@ class CookTest(unittest.TestCase):
         jobs = util.query_jobs(self.cook_url, True, job=[job_fast, job_slow]).json()
         self.assertEqual('success', jobs[0]['state'], f"Job details: {json.dumps(jobs[0], sort_keys=True)}")
         self.assertEqual('failed',  jobs[1]['state'], f"Job details: {json.dumps(jobs[1], sort_keys=True)}")
-        self.assertEqual(CookTest.REASON_CMD_NON_ZERO_EXIT, jobs[1]['instances'][0]['reason_code'])
+        self.assertEqual(reasons.CMD_NON_ZERO_EXIT, jobs[1]['instances'][0]['reason_code'])
         # Now try to kill the group again
         # (ensure it still works when there are no live jobs)
         util.kill_groups(self.cook_url, [group_uuid])

--- a/integration/tests/cook/test_basic.py
+++ b/integration/tests/cook/test_basic.py
@@ -13,7 +13,9 @@ from tests.cook import util
 class CookTest(unittest.TestCase):
     _multiprocess_can_split_ = True
 
-    CMD_NON_ZERO_EXIT_REASON = 99003
+    REASON_CMD_NON_ZERO_EXIT     = 99003
+    REASON_MAX_RUNTIME_EXCEEDED  =  2003
+    REASON_EXECUTOR_UNREGISTERED =  6002
 
     @staticmethod
     def minimal_group(**kwargs):
@@ -213,19 +215,32 @@ class CookTest(unittest.TestCase):
         job_timeout_interval_ms = job_timeout_interval_seconds * 1000
         max_runtime_ms = 5000
         assert max_runtime_ms < job_timeout_interval_ms
-        job_uuid, resp = util.submit_job(self.cook_url, command='sleep %s' % job_timeout_interval_seconds,
-                                         executor=job_executor_type, max_runtime=max_runtime_ms)
-        self.assertEqual(201, resp.status_code, msg=resp.content)
-        util.wait_for_job(self.cook_url, job_uuid, 'completed', job_timeout_interval_ms)
-        job = util.load_job(self.cook_url, job_uuid)
-        self.assertEqual(1, len(job['instances']))
-        self.assertEqual('failed', job['instances'][0]['status'])
-        self.assertEqual(2003, job['instances'][0]['reason_code'])
-        self.assertEqual('Task max runtime exceeded', job['instances'][0]['reason_string'])
-        if job_executor_type == 'cook':
-            job = util.wait_for_exit_code(self.cook_url, job_uuid)
-            self.assertEqual(-15, job['instances'][0]['exit_code'])
-            self.assertTrue(bool(job['instances'][0]['sandbox_directory']))
+        try:
+            job_uuid, resp = util.submit_job(self.cook_url,
+                                             command=f'sleep {job_timeout_interval_seconds}',
+                                             executor=job_executor_type, max_runtime=max_runtime_ms)
+            self.assertEqual(201, resp.status_code, msg=resp.content)
+            util.wait_for_job(self.cook_url, job_uuid, 'running')
+            util.wait_for_job(self.cook_url, job_uuid, 'completed', job_timeout_interval_ms)
+            job = util.wait_for_end_time(self.cook_url, job_uuid)
+            self.assertEqual(1, len(job['instances']))
+            instance = job['instances'][0]
+            self.assertEqual('failed', instance['status'])
+            valid_reasons = [# we killed the job because it took too long
+                             CookTest.REASON_MAX_RUNTIME_EXCEEDED,
+                             # the job was killed before the executor was properly registered
+                             CookTest.REASON_EXECUTOR_UNREGISTERED,
+                             # the job didn't die gracefully when we terminated it
+                             CookTest.REASON_CMD_NON_ZERO_EXIT]
+            self.assertIn(instance['reason_code'], valid_reasons)
+            actual_running_time_ms = instance['end_time'] - instance['start_time']
+            self.assertGreater(actual_running_time_ms, max_runtime_ms)
+            if job_executor_type == 'cook':
+                job = util.wait_for_exit_code(self.cook_url, job_uuid)
+                self.assertNotEqual(0, job['instances'][0]['exit_code'])
+                self.assertTrue(bool(job['instances'][0]['sandbox_directory']))
+        finally:
+            util.kill_jobs(self.cook_url, [job_uuid])
 
 
     def test_get_job(self):
@@ -790,7 +805,7 @@ class CookTest(unittest.TestCase):
         jobs = util.query_jobs(self.cook_url, True, job=[job_fast, job_slow]).json()
         self.assertEqual('success', jobs[0]['state'], f"Job details: {json.dumps(jobs[0], sort_keys=True)}")
         self.assertEqual('failed',  jobs[1]['state'], f"Job details: {json.dumps(jobs[1], sort_keys=True)}")
-        self.assertEqual(CookTest.CMD_NON_ZERO_EXIT_REASON, jobs[1]['instances'][0]['reason_code'])
+        self.assertEqual(CookTest.REASON_CMD_NON_ZERO_EXIT, jobs[1]['instances'][0]['reason_code'])
         # Now try to kill the group again
         # (ensure it still works when there are no live jobs)
         util.kill_groups(self.cook_url, [group_uuid])

--- a/integration/tests/cook/util.py
+++ b/integration/tests/cook/util.py
@@ -273,9 +273,8 @@ def wait_for_exit_code(cook_url, job_id, max_delay_ms=2000):
 
     def query():
         return query_jobs(cook_url, True, job=[job_id])
-
-    def predicate(resp):
-        job = resp.json()[0]
+    def predicate(response):
+        job = response.json()[0]
         if not job['instances']:
             logger.info(f"Job {job_id} has no instances.")
         else:
@@ -297,7 +296,8 @@ def wait_for_end_time(cook_url, job_id, max_delay_ms=2000):
     and raises an exception if the max_delay_ms wait time is exceeded.
     """
     job_id = unpack_uuid(job_id)
-    query = lambda: query_jobs(cook_url, True, job=[job_id])
+    def query():
+        return query_jobs(cook_url, True, job=[job_id])
     def predicate(response):
         job = response.json()[0]
         if not job['instances']:

--- a/integration/tests/cook/util.py
+++ b/integration/tests/cook/util.py
@@ -243,7 +243,11 @@ def all_instances_killed(response):
 
 
 def wait_for_job(cook_url, job_id, status, max_delay=120000):
-    """Wait for the given job's status to change to the specified value."""
+    """
+    Wait for the given job's status to change to the specified value.
+    Returns an up-to-date job description object on success,
+    and raises an exception if the max_delay wait time is exceeded.
+    """
     job_id = unpack_uuid(job_id)
 
     def query():
@@ -258,10 +262,12 @@ def wait_for_job(cook_url, job_id, status, max_delay=120000):
     return response.json()[0]
 
 
-def wait_for_exit_code(cook_url, job_id):
+def wait_for_exit_code(cook_url, job_id, max_delay_ms=2000):
     """
     Wait for the given job's exit_code field to appear.
     (Only supported by Cook Executor jobs.)
+    Returns an up-to-date job description object on success,
+    and raises an exception if the max_delay_ms wait time is exceeded.
     """
     job_id = unpack_uuid(job_id)
 
@@ -280,12 +286,16 @@ def wait_for_exit_code(cook_url, job_id):
                 logger.info(f"Job {job_id} instance {inst['task_id']} has exit code {inst['exit_code']}.")
                 return True
 
-    response = wait_until(query, predicate, max_wait_ms=2000, wait_interval_ms=250)
+    response = wait_until(query, predicate, max_wait_ms=max_delay_ms, wait_interval_ms=250)
     return response.json()[0]
 
 
-def wait_for_end_time(cook_url, job_id):
-    """Wait for the given job's end_time field to appear in instance 0."""
+def wait_for_end_time(cook_url, job_id, max_delay_ms=2000):
+    """
+    Wait for the given job's end_time field to appear in instance 0.
+    Returns an up-to-date job description object on success,
+    and raises an exception if the max_delay_ms wait time is exceeded.
+    """
     job_id = unpack_uuid(job_id)
     query = lambda: query_jobs(cook_url, True, job=[job_id])
     def predicate(response):
@@ -299,7 +309,7 @@ def wait_for_end_time(cook_url, job_id):
             else:
                 logger.info(f"Job {job_id} instance {inst['task_id']} has end_time {inst['end_time']}.")
                 return True
-    response = wait_until(query, predicate, max_wait_ms=2000, wait_interval_ms=250)
+    response = wait_until(query, predicate, max_wait_ms=max_delay_ms, wait_interval_ms=250)
     return response.json()[0]
 
 

--- a/integration/tests/cook/util.py
+++ b/integration/tests/cook/util.py
@@ -284,6 +284,25 @@ def wait_for_exit_code(cook_url, job_id):
     return response.json()[0]
 
 
+def wait_for_end_time(cook_url, job_id):
+    """Wait for the given job's end_time field to appear in instance 0."""
+    job_id = unpack_uuid(job_id)
+    query = lambda: query_jobs(cook_url, True, job=[job_id])
+    def predicate(response):
+        job = response.json()[0]
+        if not job['instances']:
+            logger.info(f"Job {job_id} has no instances.")
+        else:
+            inst = job['instances'][0]
+            if 'end_time' not in inst:
+                logger.info(f"Job {job_id} instance {inst['task_id']} has no end time.")
+            else:
+                logger.info(f"Job {job_id} instance {inst['task_id']} has end_time {inst['end_time']}.")
+                return True
+    response = wait_until(query, predicate, max_wait_ms=2000, wait_interval_ms=250)
+    return response.json()[0]
+
+
 def all_instances_killed(response):
     """
     Helper method used with the wait_until function.

